pgdrive/envs/marl_envs/marl_intersection.py

```diff
@@ -0,0 +1,336 @@
+                )
+            )
+            # break
+        if len(env.vehicles) == 0:
+            total_r = 0
+            print("Reset")
+            env.reset()
+    env.close()
+
+
+def _profile():
+    import time
+    env = MultiAgentIntersectionEnv({"num_agents": 16})
+    obs = env.reset()
+    start = time.time()
+    for s in range(10000):
+        o, r, d, i = env.step(env.action_space.sample())
+
+        # mask_ratio = env.scene_manager.detector_mask.get_mask_ratio()
+        # print("Mask ratio: ", mask_ratio)
+
+        if all(d.values()):
+            env.reset()
+        if (s + 1) % 100 == 0:
+            print(
+                "Finish {}/10000 simulation steps. Time elapse: {:.4f}. Average FPS: {:.4f}".format(
+                    s + 1,
+                    time.time() - start, (s + 1) / (time.time() - start)
+                )
+            )
+    print(f"(PGDriveEnvV2) Total Time Elapse: {time.time() - start}")
+
+
+def _long_run():
+    # Please refer to test_ma_Intersection_reward_done_alignment()
+    _out_of_road_penalty = 3
+    env = MultiAgentIntersectionEnv(
+        {
+            "num_agents": 32,
+            "vehicle_config": {
+                "lidar": {
+                    "num_others": 8
+                }
+            },
+            **dict(
+                out_of_road_penalty=_out_of_road_penalty,
+                crash_vehicle_penalty=1.333,
+                crash_object_penalty=11,
+                crash_vehicle_cost=13,
+                crash_object_cost=17,
+                out_of_road_cost=19,
+            )
+        }
+    )
+    try:
+        obs = env.reset()
+        assert env.observation_space.contains(obs)
+        for step in range(10000):
+            act = env.action_space.sample()
+            o, r, d, i = env.step(act)
+            if step == 0:
+                assert not any(d.values())
+
+            if any(d.values()):
+                print("Current Done: {}\nReward: {}".format(d, r))
+                for kkk, ddd in d.items():
+                    if ddd and kkk != "__all__":
+                        print("Info {}: {}\n".format(kkk, i[kkk]))
+                print("\n")
+
+            for kkk, rrr in r.items():
+                if rrr == -_out_of_road_penalty:
+                    assert d[kkk]
+
+            if (step + 1) % 200 == 0:
+                print(
+                    "{}/{} Agents: {} {}\nO: {}\nR: {}\nD: {}\nI: {}\n\n".format(
+                        step + 1, 10000, len(env.vehicles), list(env.vehicles.keys()),
+                        {k: (oo.shape, oo.mean(), oo.min(), oo.max())
+                         for k, oo in o.items()}, r, d, i
+                    )
+                )
+            if d["__all__"]:
+                print('Current step: ', step)
+                break
+    finally:
+        env.close()
+
+
+if __name__ == "__main__":
+    # _draw()
+    _vis()
+    # _vis_debug_respawn()
+    # _profiwdle()
+    # _long_run()
```

Comment:
So you have tried all these functions and they all work well? (If so, then close this comment.)